docs/proton-kafka.md

```diff
@@ -106,233 +114,326 @@ annotations:
+         topic='github_events';
+```
+
+2. Extract fields using shortcut syntax:
+
+```sql
+SELECT 
+    raw:actor AS actor,
+    raw:created_at::datetime64(3, 'UTC') AS created_at,
+    raw:id AS id,
+    raw:payload AS payload,
+    raw:repo AS repo,
+    raw:type AS type
+FROM ext_json_raw;
+```
+
+This method is most flexible and is best for dynamic JSON text with new fields or missing fields and it can also extract nested JSON fields. 
+
+#### Option B: Use `JSONEachRow` Format
+
+Define a Kafka external stream with columns which are mapped to the JSON fields and also specify the `data_format` as `JSONEachRow`. 
+
+```sql
+CREATE EXTERNAL STREAM ext_json_parsed
+    (
+        actor string,
+        created_at datetime64(3, 'UTC'),
+        id string,
+        payload string,
+        repo string,
+        type string
+    )
 SETTINGS type='kafka',
          brokers='localhost:9092',
          topic='github_events',
          data_format='JSONEachRow'
 ```
 
-If there are nested complex JSON in the message, you can define the column as a string type. Actually any JSON value can be saved in a string column.
+When users query the `ext_json_parsed` stream, the JSON fields will be parsed and casted to the target column type automatically. 
 
-:::info
+This method is most convient when the JSON text is in stable schema and can be used to extract JSON fields at top level. 
 
-Protobuf messages can be read with all or partial columns. Please check [this page](/proton-format-schema).
+### Reading CSV Kafka Messages 
 
-:::
+Similar to data format `JSONEachRow`, users can read Kafka message in CSV format.
 
-### Query Settings
+```
+CREATE EXTERNAL STREAM ext_json_parsed
+    (
+        actor string,
+        created_at datetime64(3, 'UTC'),
+        id string,
+        payload string,
+        repo string,
+        type string
+    )
+SETTINGS type='kafka',
+         brokers='localhost:9092',
+         topic='csv_topic',
+         data_format='CSV';
+```
 
-#### shards
+###  Reading TSV Kafka Messages
 
-Starting from Proton 1.3.18, you can also read in specified Kafka partitions. By default, all partitions will be read. But you can also read from a single partition via the `shards` setting, e.g.
+Identical to CSV, but expects **tab-separated values**:
 
 ```sql
-SELECT raw FROM ext_stream SETTINGS shards='0'
+SETTINGS data_format='TSV';
 ```
 
-Or you can specify a set of partition ID, separated by comma, e.g.
+### Reading Avro or Protobuf Messages
 
-```sql
-SELECT raw FROM ext_stream SETTINGS shards='0,2'
-```
+To read Avro-encoded / Protobuf-encoded Kafka message, please refer to [Schema](/proton-format-schema) and [Schema Registry](/proton-schema-registry) for details.
 
-### Read existing messages {#rewind}
+### Accessing Kafka Message Metadata
 
-When you run `SELECT raw FROM ext_stream `, Timeplus will read the new messages in the topics, not the existing ones.
+Timeplus provides **virtual columns** for Kafka message metadata.
```

Comment:
Yes, the table below is very clean but I found the docs search works better for the keywords in the heading, not so great for content in a markdown table. But it's a balance and I agree having a table is more clean for human being.